src/main/java/org/mskcc/cmo/metadb/service/impl/RequestReplyHandlingServiceImpl.java

```diff
@@ -0,0 +1,190 @@
+
+    @Value("${num.new_request_handler_threads}")
+    private int NUM_NEW_REQUEST_HANDLERS;
+
+    @Autowired
+    private CmoLabelGeneratorService cmoLabelGeneratorService;
+
+    private final ObjectMapper mapper = new ObjectMapper();
+    private static Gateway messagingGateway;
+    private static final Log LOG = LogFactory.getLog(RequestReplyHandlingServiceImpl.class);
+    private static boolean initialized = false;
+    private static volatile boolean shutdownInitiated;
+
+    private static final ExecutorService exec = Executors.newCachedThreadPool();
+    private static final BlockingQueue<ReplyInfo> cmoLabelGeneratorReqReplyQueue =
+            new LinkedBlockingQueue<ReplyInfo>();
+    private static CountDownLatch cmoLabelGeneratorHandlerShutdownLatch;
+
+
+    private class ReplyInfo {
+        String requestMessage;
+        String replyTo;
+
+        ReplyInfo(String requestMessage, String replyTo) {
+            this.requestMessage = requestMessage;
+            this.replyTo = replyTo;
+        }
+
+        String getRequestMessage() {
+            return requestMessage;
+        }
+
+        String getReplyTo() {
+            return replyTo;
+        }
+    }
+
+    private class CmoLabelGeneratorReqReplyHandler implements Runnable {
+
+        final Phaser phaser;
+        boolean interrupted = false;
+
+        CmoLabelGeneratorReqReplyHandler(Phaser phaser) {
+            this.phaser = phaser;
+        }
+
+        @Override
+        public void run() {
+            phaser.arrive();
+            while (true) {
+                try {
+                    // reply info request message contains cmo patient id
+                    ReplyInfo replyInfo = cmoLabelGeneratorReqReplyQueue.poll(100, TimeUnit.MILLISECONDS);
+                    if (replyInfo != null) {
+                        SampleMetadata sample = mapper.readValue(replyInfo.getRequestMessage(),
+                                SampleMetadata.class);
+                        String updatedCmoSampleLabel =
+                                cmoLabelGeneratorService.generateCmoSampleLabel(sample,
+                                        getExistingPatientSamples(sample.getCmoPatientId()));
+
+                        //log replied to the message
+                        messagingGateway.replyPublish(replyInfo.getReplyTo(), updatedCmoSampleLabel);
+                    }
+                    if (interrupted && cmoLabelGeneratorReqReplyQueue.isEmpty()) {
+                        break;
+                    }
+                } catch (InterruptedException e) {
+                    interrupted = true;
+                } catch (Exception e) {
+                    e.printStackTrace();
+                    LOG.error("Error during request handling", e);
+                }
+            }
+            cmoLabelGeneratorHandlerShutdownLatch.countDown();
+        }
+    }
+
+    private List<SampleMetadata> getExistingPatientSamples(String cmoPatientId) throws Exception {
```

Comment:
`MessageHandlingServiceImpl.java` uses the same method but didn't setup a common version because they `MessagingGateway`